File: Intro_Sort/introsort.py
'''
Introsort sort is a sorting algorithm that initially uses quicksort, 
but switches to heap sort if the depth of the recursion is too deep,
and uses insertion sort for small cases.
'''

#a function that takes a list as argument.
def introsort(nlist): 
    
    #maxdepth is chosen equal to 2 times floor of log base 2 of the length of the list.
    maxdepth = (len(nlist).bit_length() - 1 ) * 2
    # Call introsort_helper with start = 0 and end = len(alist).
    introsort_helper(nlist, 0, len(nlist), maxdepth)
 
#The function sorts the list from indexes start to end. 
def introsort_helper(nlist, start, end, maxdepth):
    if end - start <= 1:
        return
    elif maxdepth == 0:
        heapsort(alist, start, end)
    else:
        p = partition(nlist, start, end)
        introsort_helper(nlist, start, p + 1, maxdepth - 1)
        introsort_helper(nlist, p + 1, end, maxdepth - 1)
 
#partition function uses Hoare partition scheme to partition the list. 
def partition(nlist, start, end):
    pivot = nlist[start]
    i = start - 1
    j = end
 
    while True:
        i = i + 1
        while nlist[i] < pivot:
            i = i + 1
        j = j - 1
        while nlist[j] > pivot:
            j = j - 1
 
        if i >= j:
            return j
 
        swap(nlist, i, j)

#function for swapping 
def swap(nlist, i, j):
    nlist[i], nlist[j] = nlist[j], nlist[i]

#heapsort from indexes start to end - 1 
def heapsort(alist, start, end):
    build_max_heap(nlist, start, end)
    for i in range(end - 1, start, -1):
        swap(alist, start, i)
        max_heapify(nlist, index = 0, start = start, end = i)
 
#rearrange the list into a list representation of a heap. 
def build_max_heap(nlist, start, end):
    def parent(i):
        return (i - 1) // 2
    length = end - start
    index = parent(length - 1)
    while index >= 0:
        max_heapify(nlist, index, start, end)
        index = index - 1
 
#modifies the heap structure at and below the node at given index to make it satisfy the heap property. 
def max_heapify(nlist, index, start, end):
    def left(i):
        return 2 * i + 1
    def right(i):
        return 2 * i + 2
 
    size = end - start
    l = left(index)
    r = right(index)
    if (l < size and nlist[start + l] > nlist[start + index]):
        largest = l
    else:
        largest = index
    if (r < size and nlist[start + r] > nlist[start + largest]):
        largest = r
    if largest != index:
        swap(alist, start + largest, start + index)
        max_heapify(alist, largest, start, end)
        
#taking list as input from user (dynamic input)
nlist = input('Enter the list of numbers: ').split()
nlist = [int(n) for n in nlist]
introsort(nlist)
print('Sorted list: ', end = '')
#printing sorted list
print(nlist)


'''
Sample input
Enter the list of numbers: 8 4 3 2 1

Sample output
Sorted list: [1, 2, 3, 4, 8]
'''

File: Intro_Sort/Intro_Sort.js
//++++++++++Utility++++++++++++++++++++++

//function to swap two values
function swap(arr, index1, index2)
{
    var temp = arr[index1];
    arr[index1] = arr[index2];
    arr[index2] = temp;
}
//function to implement heapsort
function HeapSort(input)
{
    var n=input.length;
    //Build heap
    for(var i = Math.floor(n/2) - 1; i >= 0; i-- )
    {
        heapify( arr, n, i);
    }

    for( var i = n - 1; i >= 0; i-- )
    {
        swap( arr, 0 , i);
        heapify ( arr ,i, 0);
    }
}
//This function maintain property of max heap
function heapify( arr , n, i)
{
    var greatest = i;
    var idx_left_child = 2*i + 1;
    var idx_right_child = 2*i + 2;
    //If left child child is greater then its parent
    if ( ( idx_left_child  <  n )   &&   ( arr[idx_left_child]  >  arr[greatest] ) )
    {
        greatest = idx_left_child;
    }
    //If right child is greater then its parent
    if ( ( idx_right_child  <  n )   &&   ( arr[idx_right_child]  >  arr[greatest] ) )
    {
        greatest = idx_right_child;
    }
    //One of the above condition was true then heap property is not being satisfied 
    //Hence we heapify the sub-tree whose root is at ith index of array
    if(greatest != i)
    {
        swap(arr, i, greatest);
        heapify( arr, n, greatest);
    }
}
//Function that performs Insertion Sort
function InsertionSort( nums, begin, end)
{
    for (var i = 1; i < nums.length; i++)
    {
        let j = i - 1;
        let temp = nums[i];
        while (j >= 0 && nums[j] > temp) 
        {
          nums[j + 1] = nums[j];
          j--;
        }
        nums[j+1] = temp;
    }
}
//Function that finds middle of the numbers a, b, c 
function MedianofThree(a, b, c)
{
    if(a < b && b < c)
        return b;
    if( a < c && c <= b)
        return c;
    if(b <= a && a < c)
        return a;
    if( b < c && c <= a)
        return c;
    if( c <= a && a < b)
        return a;
    if( c <= b && b <= a)
        return b;
}
//Function to create partition for Quick Sort
function Part(arr, begin,end)
{
    var pivot=arr[Math.floor((begin + end) / 2)];
    var i = begin;
    var j = end;
    while (i <= j)
    {
        while( arr[i] < pivot)
        {
            i++;
        }
        while( arr[j] > pivot)
        {
            j--;
        }
        if( i<= j)
        {
            swap(arr, i, j);
            i++;
            j--;
        }
    }
    return i;
}

//+++++++++++++++End of Utilities++++++++++++++++++++++++++++++

function IntroSort( arr )
{
    var depth = 2 * Math.floor(  Math.log( arr.length )  );
    IntroSortHelper( arr, 0, arr.length-1, depth );
    return ;
}
//Function that Implement IntroSort
function IntroSortHelper( arr, begin, end, depth)
{
    var size = end - begin;
    //If depth is zero then heap sort is used
    if ( depth === 0 )
    {
        //Below commented statement can be used to check the working of the inputs provided
        //console.log('Depth is zero hence using Heap Sort:\n');
        HeapSort(arr);
        return ;
    }
    //If number of elements is less then 16 Insertion Sort is used
    else if (size < 16)
    {
        //Below commented statement can be used to check the working of the inputs provided
        //console.log('Number of elements less then 16 hence using Insertion Sort:\n');
        InsertionSort (arr, begin, end);
        return ;
    }
    else
    {
        //Below commented statement can be used to check the working of the inputs provided
        //console.log('Number of elements greater then 16 hence using Quick Sort: ')
        var pivot = MedianofThree( begin, Math.floor(begin+size/2), end);
        swap(arr, pivot, end);
        var partition = Part(arr, begin , end);
        IntroSortHelper(arr, begin, partition-1, depth-1);
        IntroSortHelper(arr, partition+1, end , depth-1);
        return ;
    }
}
//++++++++++++++++++++++++++Input+++++++++++++++++++++++++++++++++++++++++
var readlineSync = require('readline-sync');
var n = readlineSync.question('Enter the number of elements: ')
var arr = [n];
for( var i = 0; i < n ; i++ )  
{
    console.log(`Enter the `, i ,`th element of array: `);
    var a = readlineSync.question();
    arr[i] = a;
}
IntroSort(arr);
console.log(arr);

/*
1. When initially Quick Sort is used

Input1: [2, 4, 7, 89, 0, 73, 3, 4, -2, -8, 22, 28, 19, 100, -3333, -89, 200098, -10]
Output1: [-3333, -89, -10, -8, -2, 0, 2,  3, 4, 4,   7, 19, 22, 28,  73, 89, 100, 200098]


2.When Insertion Sort is used

Input2: [3, 7, 2, 9, 1]
Output2: [3, 7, 2, 9, 1]


3. When Heap Sort is used

Input3: [9,8]
Output3: [8,9]
*/
